File: buzzad-benefit/app/src/main/java/com/buzzvil/buzzad/benefit/sample/publisher/nativead2/carousel/NativeAd2CarouselViewHolder.java
package com.buzzvil.buzzad.benefit.sample.publisher.nativead2.carousel;

import android.view.View;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.buzzvil.buzzad.benefit.core.ad.AdError;
import com.buzzvil.buzzad.benefit.nativead2.api.NativeAd2;
import com.buzzvil.buzzad.benefit.nativead2.api.NativeAd2EventListener;
import com.buzzvil.buzzad.benefit.nativead2.api.NativeAd2Pool;
import com.buzzvil.buzzad.benefit.nativead2.api.NativeAd2StateChangedListener;
import com.buzzvil.buzzad.benefit.nativead2.api.NativeAd2View;
import com.buzzvil.buzzad.benefit.nativead2.api.NativeAd2ViewBinder;
import com.buzzvil.buzzad.benefit.presentation.feed.entrypoint.FeedPromotion;
import com.buzzvil.buzzad.benefit.presentation.feed.entrypoint.FeedPromotionViewBinder;
import com.buzzvil.buzzad.benefit.presentation.media.DefaultCtaView;
import com.buzzvil.buzzad.benefit.presentation.media.MediaView;
import com.buzzvil.buzzad.benefit.presentation.reward.RewardResult;
import com.buzzvil.buzzad.benefit.sample.publisher.R;

public class NativeAd2CarouselViewHolder extends RecyclerView.ViewHolder {

    private TextView positionTextView;
    private LinearLayout loadingView;

    private NativeAd2ViewBinder nativeAd2ViewBinder;
    private FeedPromotionViewBinder carouselToFeedViewBinder;

    NativeAd2CarouselViewHolder(String unitId, NativeAd2View nativeAd2View) {
        super(nativeAd2View);

        positionTextView = nativeAd2View.findViewById(R.id.positionTextView);
        TextView stateTextView = nativeAd2View.findViewById(R.id.stateTextView);
        TextView eventTextView = nativeAd2View.findViewById(R.id.eventTextView);
        this.loadingView = nativeAd2View.findViewById(R.id.loadingView);

        MediaView mediaView = nativeAd2View.findViewById(R.id.mediaView);
        ImageView imageIcon = nativeAd2View.findViewById(R.id.imageIcon);
        TextView textTitle = nativeAd2View.findViewById(R.id.textTitle);
        TextView textDescription = nativeAd2View.findViewById(R.id.textDescription);
        DefaultCtaView ctaView = nativeAd2View.findViewById(R.id.ctaView);

        // NativeAd2View와 하위 컴포넌트를 바인드합니다.
        nativeAd2ViewBinder = new NativeAd2ViewBinder.Builder()
                .nativeAd2View(nativeAd2View)
                .mediaView(mediaView)
                .iconImageView(imageIcon)
                .titleTextView(textTitle)
                .descriptionTextView(textDescription)
                .ctaView(ctaView)
                .build(unitId);

        // 필요에 따라 광고 갱신 상태를 받아올 수 있는 NativeAd2StateChangedListener를 설정하세요.
        nativeAd2ViewBinder.addNativeAd2StateChangedListener(new NativeAd2StateChangedListener() {
            @Override
            public void onRequested() {
                stateTextView.setText("state: onRequested");
            }

            @Override
            public void onNext(@NonNull NativeAd2 nativeAd2) {
                stateTextView.setText("state: onNext");
            }

            @Override
            public void onComplete() {
                stateTextView.setText("state: onComplete");
            }

            @Override
            public void onError(@NonNull AdError adError) {
                stateTextView.setText("state: onError(" + adError.getAdErrorType().name() + ")");
            }
        });

        // 필요에 따라 광고 이벤트 상태를 받아올 수 있는 NativeAd2EventListener를 설정하세요.
        nativeAd2ViewBinder.addNativeAd2EventListener(new NativeAd2EventListener() {
            @Override
            public void onImpressed(@NonNull NativeAd2 nativeAd2) {
                eventTextView.setText("event: onImpressed");
            }

            @Override
            public void onClicked(@NonNull NativeAd2 nativeAd2) {
                eventTextView.setText("event: onClicked");
            }

            @Override
            public void onRewardRequested(@NonNull NativeAd2 nativeAd2) {
                eventTextView.setText("event: onRewardRequested");
            }

            @Override
            public void onRewarded(@NonNull NativeAd2 nativeAd2, @NonNull RewardResult rewardResult) {
                eventTextView.setText("event: onRewarded, result: " + rewardResult.name());
            }

            @Override
            public void onParticipated(@NonNull NativeAd2 nativeAd2) {
                eventTextView.setText("event: onParticipated");
            }
        });

        // CarouselToFeedSlideItem을 위한 바인더를 빌드합니다.
        carouselToFeedViewBinder = new FeedPromotionViewBinder.Builder(nativeAd2View, mediaView)
                .titleTextView(textTitle)
                .descriptionTextView(textDescription)
                .iconImageView(imageIcon)
                .ctaView(ctaView)
                .build();
    }

    public void setPool(NativeAd2Pool carouselPool, int adKey) {
        // 해당 position(adKey)에 해당하는 NativeAd2ViewBinder가 carouselPool을 사용하도록 합니다.
        nativeAd2ViewBinder.setPool(adKey, carouselPool);
    }

    public void bind(int position) {
        positionTextView.setText("position: " + position);
        setLoadingView(false);

        // NativeAd2ViewBinder의 bind()를 호출하면 광고 할당 및 갱신이 자동으로 수행됩니다.
        nativeAd2ViewBinder.bind(position);
    }

    public void bind(int position, FeedPromotion feedPromotion) {
        positionTextView.setText("position: " + position);
        setLoadingView(false);

        // FeedPromotion 객체를 뷰에 바인드합니다.
        carouselToFeedViewBinder.bind(feedPromotion);
    }

    public void unbind(int position) {
        // unbind를 반드시 호출하여 뷰를 재사용할 때 문제가 발생하지 않게 합니다.
        nativeAd2ViewBinder.unbind(position);
        carouselToFeedViewBinder.unbind();
    }

    private void setLoadingView(boolean isLoading) {
        if (isLoading) {
            loadingView.setVisibility(View.VISIBLE);
        } else {
            loadingView.setVisibility(View.GONE);
        }
    }
}

File: buzzvil-sdk-android/app/src/main/java/com/buzzvil/sample/buzzvilsdksample/nativead/carousel/NativeAd2CarouselViewHolder.kt
package com.buzzvil.sample.buzzvilsdksample.nativead.carousel

import android.view.View
import androidx.recyclerview.widget.RecyclerView
import com.buzzvil.buzzad.benefit.core.ad.AdError
import com.buzzvil.buzzad.benefit.nativead2.api.NativeAd2
import com.buzzvil.buzzad.benefit.nativead2.api.NativeAd2EventListener
import com.buzzvil.buzzad.benefit.nativead2.api.NativeAd2Pool
import com.buzzvil.buzzad.benefit.nativead2.api.NativeAd2StateChangedListener
import com.buzzvil.buzzad.benefit.nativead2.api.NativeAd2ViewBinder
import com.buzzvil.buzzad.benefit.presentation.feed.entrypoint.FeedPromotion
import com.buzzvil.buzzad.benefit.presentation.feed.entrypoint.FeedPromotionViewBinder
import com.buzzvil.buzzad.benefit.presentation.reward.RewardResult
import com.buzzvil.sample.buzzvilsdksample.Constant
import com.buzzvil.sample.buzzvilsdksample.databinding.CarouselItemBinding

class NativeAd2CarouselViewHolder(
    private val binding: CarouselItemBinding
) : RecyclerView.ViewHolder(binding.root) {
    private val nativeAd2ViewBinder: NativeAd2ViewBinder
    private val carouselToFeedViewBinder: FeedPromotionViewBinder

    init {
        val nativeAd2View = binding.nativeAdView
        val mediaView = binding.adMediaView
        val titleTextView = binding.adTitleText
        val descriptionTextView = binding.adDescriptionText
        val iconImageView = binding.adIconImage
        val ctaView = binding.adCtaView

        // NativeAd2View와 하위 컴포넌트를 바인드합니다.
        nativeAd2ViewBinder = NativeAd2ViewBinder.Builder()
            .nativeAd2View(nativeAd2View)
            .mediaView(mediaView)
            .titleTextView(titleTextView)
            .descriptionTextView(descriptionTextView)
            .iconImageView(iconImageView)
            .ctaView(ctaView)
            .build(Constant.YOUR_NATIVE_ID)

        // 광고 로딩 상태에 따라 로딩 뷰를 보여주거나 숨깁니다.
        nativeAd2ViewBinder.addNativeAd2StateChangedListener(object : NativeAd2StateChangedListener {
            override fun onComplete() {
                setLoadingView(false)
            }

            override fun onError(adError: AdError) {
                setLoadingView(false)
            }

            override fun onNext(nativeAd2: NativeAd2) {
                setLoadingView(false)
            }

            override fun onRequested() {
                setLoadingView(true)
            }
        })

        // 필요에 따라 광고 이벤트 상태를 받아올 수 있는 NativeAd2EventListener를 설정하세요.
        // 로그 기록, 단순 알림 외에 다른 동작을 추가하는 것을 권장하지 않습니다. 자동 갱신 등 네이티브 2.0의 기능과 직접 구현한 동작이 충돌할 수 있습니다.
        nativeAd2ViewBinder.addNativeAd2EventListener(object : NativeAd2EventListener {
            override fun onClicked(nativeAd2: NativeAd2) {}
            override fun onImpressed(nativeAd2: NativeAd2) {}
            override fun onParticipated(nativeAd2: NativeAd2) {}
            override fun onRewardRequested(nativeAd2: NativeAd2) {}
            override fun onRewarded(nativeAd2: NativeAd2, rewardResult: RewardResult) {}
        })

        // CarouselToFeedSlideItem을 위한 바인더를 빌드합니다.
        carouselToFeedViewBinder = FeedPromotionViewBinder.Builder(nativeAd2View, mediaView)
            .titleTextView(titleTextView)
            .descriptionTextView(descriptionTextView)
            .iconImageView(iconImageView)
            .ctaView(ctaView)
            .build()
    }

    fun setPool(carouselPool: NativeAd2Pool, adKey: Int) {
        // 해당 position(adKey)에 해당하는 NativeAd2ViewBinder가 carouselPool을 사용하도록 합니다.
        nativeAd2ViewBinder.setPool(adKey, carouselPool)
    }

    fun bind(position: Int) {
        // NativeAd2ViewBinder의 bind()를 호출하면 광고 할당 및 갱신이 자동으로 수행됩니다.
        nativeAd2ViewBinder.bind(position)
    }

    fun bind(position: Int, feedPromotion: FeedPromotion) {
        // FeedPromotion 객체를 뷰에 바인드합니다.
        carouselToFeedViewBinder.bind(feedPromotion)

        // 베네핏허브 진입 슬라이드는 로딩 뷰가 필요 없기 때문에 false로 설정합니다.
        setLoadingView(false)
    }

    fun unbind(position: Int) {
        // unbind를 반드시 호출하여 뷰를 재사용할 때 문제가 발생하지 않게 합니다.
        nativeAd2ViewBinder.unbind(position)
        carouselToFeedViewBinder.unbind()
    }

    private fun setLoadingView(isLoading: Boolean) {
        if (isLoading) {
            binding.loadingView.visibility = View.VISIBLE
        } else {
            binding.loadingView.visibility = View.GONE
        }
    }
}
